File: Sorting Algorithms/BubbleSort.java
import java.util.Arrays;

public class BubbleSort {
    public static void main(String[] args) {
        int[] input = {5,1,4,2,8};
        System.out.println("Unsorted Array = " + Arrays.toString(input));
        BubbleSort(input);
        System.out.println("Sorted Array = " + Arrays.toString(input) + " (Using Bubble Sort)");
    }
    /*
     * Bubble Sort is the simplest sorting algorithm that works by repeatedly swapping
     *  the adjacent elements if they are in the wrong order.
     * Worst Case Time Complexity [ Big-O ]: O(n^2) 
     * Best Case Time Complexity [Big-omega]: O(n) (When the list is already sorted)
     * Average Time Complexity [Big-theta]: O(n^2) 
     * Space Complexity: O(1) (Only a single additional memory space is required i.e. for temp variable.)
     */
    public static int[] BubbleSort(int[] array){
        int n = array.length - 1;
        boolean swapped = true;
        int temp;
        while(swapped){
            swapped = false;
            for(int i=0; i<n; i++){
                if(array[i]> array[i+1]){
                    temp = array[i];
                    array[i] = array[i+1];
                    array[i+1] = temp;
                    swapped = true;
                }
            }
            n--;
        }
        return array;
    }
}

File: Sorting Algorithms/C++/BubbleSort.cpp
#include <iostream>

void printVector(std::vector<int> &vector){
    for(int i=0; i<vector.size(); i++){
        if(i==0){
            std::cout << "[";
        }
        if(i== vector.size()-1){
            std::cout << vector[i] << "]" << std::endl;
        } else {
            std::cout << vector[i] << ",";
        }
    }
}

void swap(std::vector<int> &vector, int i, int j){
    int temp = vector[j];
    vector[j] = vector[i];
    vector[i] = temp;
}

/*
Time O(n^2) | Space O(1)
Stable
Adaptive
*/
std::vector<int> bubbleSort(std::vector<int> &vector){
    bool hasSwapped = false;
    for(int i=0; i<vector.size(); i++){
        hasSwapped = true;
        for(int j=0; j<vector.size()-1-i;j++){
            if(vector[j]>vector[j+1]){
                swap(vector,j,j+1);
                hasSwapped = true;
            }
        }
//        std::cout << "Pass " << i << " = ";
//        printVector(vector);
        if(hasSwapped == false){
            return vector;
        }
    }
    return vector;
}

int main(int argc, char const *argv[])
{
    std::vector<int> testCase = {1,2,5,-2,10};
    std::cout << "Input = ";
    printVector(testCase);

    std::vector<int> result = bubbleSort(testCase);
    std::cout << "Result = ";
    printVector(result);

    return 0;
}

